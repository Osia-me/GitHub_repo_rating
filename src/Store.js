import React, { useState, useEffect } from "react";
import axios from 'axios';

export const Context = React.createContext(null);

export const Store = ({children}) => {
  const [state, setState] = useState([]);
  const linkToLastWeek = "https://api.github.com/search/repositories?q=created:%3E2021-06-13&sort=stars&order=desc";
  const linkToReposByLang = (lang) => `https://api.github.com/legacy/repos/search/language=${lang}`

  useEffect(() => {
    axios
      .get(linkToLastWeek)
      .then((response) => {
        setState({
          items: response.data.items,
          fav: [],
          view: null,
          language: 'Select language'
        });
      })
      .catch((error) => {
        alert('Something went wrong!')
      });
  }, []);

  useEffect(() => {
    if(state.language && state.language !== 'Select language'){
      axios.get(linkToReposByLang(state.language)).then((response) => {
        setState({
          items: response.data.repositories.map( el => ({
            name: el.name,
            description:el.description,
            html_url: el.url,
            stargazers_count: Math.round(el.score),
            id: Math.random() * 2.5
          })),
          fav: [],
          view: null,
          language: state.language
        });
      })
      .catch((error) => {
        alert('Something went wrong!')
      });
    }
  },[state.language])

  return (
    <Context.Provider value={[state, setState]}>
      { children }
    </Context.Provider>
  );
};

export default Store;
