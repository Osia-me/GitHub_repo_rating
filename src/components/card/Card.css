.card {
  flex-direction: column;
  position: relative;
  background-color: #2c333e;
  width: 30%;
  height: 13em;
  border-color: red;
  margin: 15px 1em;
  border-radius: 5px;
  font-family: inherit;
  font-size: 1em;
}

.card-text {
  width: 100%;
  text-align: center;
}

.card-name {
  color: #74dc16;
  align-items: center;
  border: 1px solid #0d1117;
  margin: 1em 1em .2em 1em;
  padding: .3em;
  border-radius: 3px;
  height: 2em;
}

.card-info {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
}

.card-decsription {
  color: #08a08b;
  align-items: center;
  text-align: left;
  border: 1px solid #0d1117;
  margin: .2em 1em;
  padding: 0.2em .5em;
  border-radius: 3px;
  width: 80%;
  height: 7em;
}

.star {
  margin-right: 1em;
}

.star-number {
  color: #D5C90C;
}

a {
    color: #de4fda;
    text-decoration: none;
}

a:hover {
    color:#00A0C6;
    text-decoration:none;
    cursor:pointer;
}

.card-link {
  text-align: center;
  position: absolute;
  left: 50%;
  transform: translateX(-50%);
  bottom: .5em;
}

@media only screen and (max-width: 920px) {
  .card {
    width: 100%;
    border-color: red;
    margin: 15px .5em;
  }
}
