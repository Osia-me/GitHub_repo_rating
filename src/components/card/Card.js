import React, { useState, useEffect, useContext } from 'react';
import Star from '../elements/Star/Star';
import { Context } from '../../Store.js';

import './Card.css';

function Card({id, name, description, numberOfStars, link}) {
  const [state, setState] = useContext(Context);
  const [updatedNumberOfStars, setUpdatedNumberOfStars] = useState(numberOfStars);
  const [clickStar, setClickStar] = useState(0);
  const [updateDescription, setUpdateDescription] = useState(description);

  const containJapanese = (string) => string.match(/[\u3400-\u9FBF]/);

  const StarClickHandlerOnCard = () => {

    clickStar === 0 ? setClickStar(1) : setClickStar(0);

    let output = AddOrDeleteId(state.fav.map(el => el.id), id);
    setState({
      ...state,
      fav: state.items.filter(value => output.includes(value.id))
    });
  }

  const AddOrDeleteId = (arr, id) => {
    if(arr.includes(id)){
      arr = arr.filter(inp => inp !== id);
    } else {
      arr.push(id);
    }
    return arr;
  }

  useEffect(() => {
    const updatedNumber = numberOfStars +1;
    if(clickStar === 1){
      setUpdatedNumberOfStars(updatedNumber);
    } else {
      setUpdatedNumberOfStars(numberOfStars);
    }
  },[clickStar, numberOfStars]);

  useEffect(() => {
    if(description && description.length > 80 && containJapanese(description))
      setUpdateDescription(`${description.substring(0,28)}...`);
    if(description && description.length > 80 && !containJapanese(description))
      setUpdateDescription(`${description.substring(0,80)}...`);
    if(!description || description === '')
      setUpdateDescription("This repository doesn't contain any description")
  },[description]);

  return (
    <div className='card'>
      <div className='card-text'>
        <div className='card-name'>Name: { name.length > 20 ? `${name.substring(0,20)}...` : name}</div>
        <div className='card-info'>
        <div className='card-decsription'>{updateDescription}</div>
        <div className='star'>
          <Star width={25} height={25} starclick={StarClickHandlerOnCard} color = {clickStar ? 'color' : null}/>
          <span className='star-number'>
            {updatedNumberOfStars}
          </span>
        </div>
        </div>
        <div className='card-link'>
          <a href={link} target="_blank" rel="noreferrer">Visit it</a>
        </div>
      </div>
    </div>
  );
}

export default Card;
