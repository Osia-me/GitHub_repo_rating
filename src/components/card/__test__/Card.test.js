import Store from '../../../Store.js';
import { shallow } from 'enzyme';
import Card from "../Card";

test('renders Card without crashing', () => {
  const wrapper = shallow(
    <Store className="app">
      <Card/>
    </Store>);

  expect(wrapper).toMatchSnapshot();
  expect(wrapper.contains(<Card />)).toBe(true);
});
