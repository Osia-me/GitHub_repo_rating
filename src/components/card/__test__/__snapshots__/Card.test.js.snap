// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Card without crashing 1`] = `ShallowWrapper {}`;
